cc/gtp_player.cc

```diff
@@ -0,0 +1,323 @@
+    return HandleGenmove(cmd, args);
+  } else if (cmd == "play") {
+    return HandlePlay(cmd, args);
+  } else if (cmd == "readouts") {
+    return HandleReadouts(cmd, args);
+  } else if (cmd == "report_search_interval") {
+    return HandleReportSearchInterval(cmd, args);
+  }
+  std::cerr << "ERROR: unrecognized GTP command \"" << cmd << "\"\n";
+  return "";
+}
+
+std::string GtpPlayer::HandleClearBoard(
+    absl::string_view cmd, const std::vector<absl::string_view>& args) {
+  if (!CheckArgsExact(cmd, 0, args)) {
+    return "";
+  }
+
+  NewGame();
+
+  return "";
+}
+
+std::string GtpPlayer::HandleEcho(absl::string_view cmd,
+                                  const std::vector<absl::string_view>& args) {
+  return absl::StrJoin(args, " ");
+}
+
+std::string GtpPlayer::HandleFinalScore(
+    absl::string_view cmd, const std::vector<absl::string_view>& args) {
+  if (!CheckArgsExact(cmd, 0, args)) {
+    return "";
+  }
+  return absl::StrJoin(args, " ");
+}
+
+std::string GtpPlayer::HandleGamestate(
+    absl::string_view cmd, const std::vector<absl::string_view>& args) {
+  if (!CheckArgsExact(cmd, 0, args)) {
+    return "";
+  }
+
+  const auto& position = root()->position;
+
+  // board field.
+  std::ostringstream oss;
+  for (const auto& stone : position.stones()) {
+    char ch;
+    if (stone.color() == Color::kBlack) {
+      ch = 'X';
+    } else if (stone.color() == Color::kWhite) {
+      ch = 'O';
+    } else {
+      ch = '.';
+    }
+    oss << ch;
+  }
+  std::string board = oss.str();
+
+  // toPlay field.
+  std::string to_play = position.to_play() == Color::kBlack ? "Black" : "White";
+
+  // lastMove field.
+  std::string last_move;
+  if (!history().empty()) {
+    last_move = absl::StrCat("\"", history().back().c.ToKgs(), "\"");
+  } else {
+    last_move = "null";
+  }
+
+  // n field.
+  auto n = position.n();
+
+  // q field.
+  auto q = root()->parent != nullptr ? root()->parent->Q() : 0;
+
+  // clang-format off
+    std::cerr << "mg-gamestate: {"
+              << "\"board\":\"" << board << "\", "
+              << "\"toPlay\":\"" << to_play << "\", "
+              << "\"lastMove\":" << last_move << ", "
+              << "\"n\":" << n << ", "
+              << "\"q\":" << q
+              << "}" << std::endl;
+  // clang-format on
+  return "";
+}
+
+std::string GtpPlayer::HandleGenmove(
+    absl::string_view cmd, const std::vector<absl::string_view>& args) {
+  if (!CheckArgsRange(cmd, 0, 1, args)) {
+    return "";
+  }
+
+  auto c = SuggestMove(num_readouts_);
+  std::cerr << root()->Describe() << std::endl;
+  if (ShouldResign()) {
+    return "resign";
+  }
+  PlayMove(c);
+
+  return c.ToKgs();
+}
+
+std::string GtpPlayer::HandlePlay(absl::string_view cmd,
+                                  const std::vector<absl::string_view>& args) {
+  if (!CheckArgsExact(cmd, 2, args)) {
+    return "";
+  }
+
+  Color color;
+  if (args[0] == "b" || args[0] == "B") {
```

Comment:
I'm realizing that maybe you don't want a pedantic review so I'll stop after this comment

http://www.lysator.liu.se/~gunnar/gtp/gtp2-spec-draft2/gtp2-spec.html
3.2.1 says color can be any of {"b", "w", "black", "white"} case insensitive

Reply:
updated to accept anything that begins with 'b' or 'w' :P